### PR TITLE
Add palette categories

### LINKED ISSUE #496: Palette Categories
The palette should be separated into separate categories like "Image Processing", "Feature Detection", etc...

### PR DESCRIPTION
**Pull request by ThomasJClark**
This should make it a lot easier to find operations, and it means we don't have to add operations in a certain order anymore to force them into categories.

palette

Closes #496

(Again, formatting changes were done weeks ago and I'm not gonna undo them)

### CODE REVIEW

File: core/src/main/java/edu/wpi/grip/core/Operation.java [hunk 1]
```diff
@@ -23,6 +31,13 @@
     String getDescription();
 
     /**
+     * @return What category the operation falls under.  This is used to organize them in the GUI
+     */
+    default Category getCategory() {
+        return Category.MISCELLANEOUS;
+    }
```
Comment (JLLeitschuh) [inline core/src/main/java/edu/wpi/grip/core/Operation.java]: Why have a default?

Comment (ThomasJClark) [inline core/src/main/java/edu/wpi/grip/core/Operation.java]: I can actually remove this.  I added it at first, but I think we generally want to specify it explicitly.

**Comment by JLLeitschuh** [comment]
I'm not done reviewing this. I'll finish looking on Monday

File: core/src/main/java/edu/wpi/grip/core/Operation.java [hunk 2]
```diff
@@ -11,6 +11,14 @@
  */
 public interface Operation {
 
+    enum Category {
```
Comment (JLLeitschuh) [inline core/src/main/java/edu/wpi/grip/core/Operation.java]: Document this.

File: core/src/main/java/edu/wpi/grip/core/operations/Operations.java [hunk 3]
```diff
@@ -33,9 +33,9 @@ public static void addOperations(EventBus eventBus) {
         eventBus.post(new OperationAddedEvent(new FilterLinesOperation()));
         eventBus.post(new OperationAddedEvent(new MaskOperation()));
         eventBus.post(new OperationAddedEvent(new MinMaxLoc()));
+        eventBus.post(new OperationAddedEvent(new MatFieldAccessor()));
```
Comment (JLLeitschuh) [inline core/src/main/java/edu/wpi/grip/core/operations/Operations.java]: Why was this file changed?

File: ui/src/main/java/edu/wpi/grip/ui/PaletteController.java [hunk 4]
```diff
@@ -58,40 +46,23 @@ public void initialize() {
...
-                        .setVisible(SearchUtility.fuzzyContains(operation.getName(), searchText)
-                                || SearchUtility.fuzzyContains(operation.getDescription(), searchText));
-            });
-        };
-
-        // Update the visibility of each node when either the list of nodes or the filter text changes
-        operations.getChildren().addListener(filterOperations);
-        operationSearch.textProperty().addListener(filterOperations);
+        // Bind the filterText of all of the individual tabs to the search field
+        operationSearch.textProperty().addListener(observable -> {
```
Comment (JLLeitschuh) [inline ui/src/main/java/edu/wpi/grip/ui/PaletteController.java]: >  Codacy Issue found: [The String literal "filterText" appears 6 times in this file; the first occurrence is on line 57](https://www.codacy.com/app/WPIRoboticsProjects/GRIP/file/2047798628/issues/source?bid=2943107&fileBranchId=3130834#l57)

Fix this please.

File: ui/src/main/java/edu/wpi/grip/ui/OperationListController.java [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+public class OperationListController {
+
+    @FXML private Tab root;
+    @FXML private VBox operations;
+
+    @Inject private OperationController.Factory operationControllerFactory;
+    @Inject private Logger logger;
+
+    private final StringProperty filterText = new SimpleStringProperty(this, "filterText", "");
+    private String baseText = null;
```
Comment (JLLeitschuh) [inline ui/src/main/java/edu/wpi/grip/ui/OperationListController.java]: > Codacy Issue found: [Perhaps 'baseText' could be replaced by a local variable.](https://www.codacy.com/app/WPIRoboticsProjects/GRIP/file/2047798627/issues/source?bid=2943107&fileBranchId=3130834#l38)

Please resolve this.

File: ui/src/main/java/edu/wpi/grip/ui/OperationListController.java [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+ * @see edu.wpi.grip.core.Operation.Category
+ */
+@ParametrizedController(url = "OperationList.fxml")
+public class OperationListController {
+
+    @FXML private Tab root;
+    @FXML private VBox operations;
+
+    @Inject private OperationController.Factory operationControllerFactory;
+    @Inject private Logger logger;
```
Comment (JLLeitschuh) [inline ui/src/main/java/edu/wpi/grip/ui/OperationListController.java]: > Codacy Issue found: [Avoid unused private fields such as 'logger'.](https://www.codacy.com/app/WPIRoboticsProjects/GRIP/file/2047798627/issues/source?bid=2943107&fileBranchId=3130834#l35)

Please resolve this

File: ui/src/main/java/edu/wpi/grip/ui/PaletteController.java [hunk 7]
```diff
@@ -24,30 +18,24 @@
...
+    @FXML private CustomTextField operationSearch;
+    @FXML private Tab allOperations;
+    @FXML private Tab imgprocOperations;
+    @FXML private Tab featureOperations;
+    @FXML private Tab networkOperations;
+    @FXML private Tab opencvOperations;
+    @FXML private Tab miscellaneousOperations;
+
+    @Inject private EventBus eventBus;
+    @Inject private Palette palette;
```
Comment (JLLeitschuh) [inline ui/src/main/java/edu/wpi/grip/ui/PaletteController.java]: > Codacy Issue found: [Avoid unused private fields such as 'palette'.](https://www.codacy.com/app/WPIRoboticsProjects/GRIP/file/2047798628/issues/source?bid=2943107&fileBranchId=3130834#l35)

Please resolve this.

File: ui/src/main/java/edu/wpi/grip/ui/PaletteController.java [hunk 8]
```diff
@@ -24,30 +18,24 @@
...
+    @FXML private VBox root;
+    @FXML private CustomTextField operationSearch;
+    @FXML private Tab allOperations;
+    @FXML private Tab imgprocOperations;
+    @FXML private Tab featureOperations;
+    @FXML private Tab networkOperations;
+    @FXML private Tab opencvOperations;
+    @FXML private Tab miscellaneousOperations;
+
+    @Inject private EventBus eventBus;
```
Comment (JLLeitschuh) [inline ui/src/main/java/edu/wpi/grip/ui/PaletteController.java]: > Codacy Issue found: [Avoid unused private fields such as 'eventBus'.](https://www.codacy.com/app/WPIRoboticsProjects/GRIP/file/2047798628/issues/source?bid=2943107&fileBranchId=3130834#l34)

Please resolve this.

File: core/src/main/java/edu/wpi/grip/core/Operation.java (hunk 1 above)
Comment (JLLeitschuh) [inline core/src/main/java/edu/wpi/grip/core/Operation.java]: Would this make more sense as an `OTHER` instead of `MISCELLANEOUS`

File: ui/src/main/java/edu/wpi/grip/ui/PaletteController.java [hunk 9]
```diff
@@ -24,30 +18,24 @@
...
-    private Palette palette;
-    private ControllerMap<OperationController, Node> operationsMapManager;
+    @FXML private VBox root;
+    @FXML private CustomTextField operationSearch;
+    @FXML private Tab allOperations;
+    @FXML private Tab imgprocOperations;
+    @FXML private Tab featureOperations;
+    @FXML private Tab networkOperations;
+    @FXML private Tab opencvOperations;
+    @FXML private Tab miscellaneousOperations;
```
Comment (JLLeitschuh) [inline ui/src/main/java/edu/wpi/grip/ui/PaletteController.java]: Remove this commit noise.

File: ui/src/main/java/edu/wpi/grip/ui/PaletteController.java (hunk 8 above)
Comment (ThomasJClark) [inline ui/src/main/java/edu/wpi/grip/ui/PaletteController.java]: k

File: ui/src/main/java/edu/wpi/grip/ui/PaletteController.java (hunk 7 above)
Comment (ThomasJClark) [inline ui/src/main/java/edu/wpi/grip/ui/PaletteController.java]: k

File: ui/src/main/java/edu/wpi/grip/ui/OperationListController.java (hunk 5 above)
Comment (ThomasJClark) [inline ui/src/main/java/edu/wpi/grip/ui/OperationListController.java]: Did you actually read the code, or are you just echoing back everything the bot says?  Because there's not way to do this as far as I can tell

File: ui/src/main/java/edu/wpi/grip/ui/OperationListController.java (hunk 6 above)
Comment (ThomasJClark) [inline ui/src/main/java/edu/wpi/grip/ui/OperationListController.java]: k

File: core/src/main/java/edu/wpi/grip/core/Operation.java (hunk 1 above)
Comment (ThomasJClark) [inline core/src/main/java/edu/wpi/grip/core/Operation.java]: Probably not.  "Other" kind of sounds like there are other categories besides the one listed, which isn't true.

File: ui/src/main/java/edu/wpi/grip/ui/OperationListController.java (hunk 5 above)
Comment (JLLeitschuh) [inline ui/src/main/java/edu/wpi/grip/ui/OperationListController.java]: The scope of the field is greater than it needs to be. You can make the scope of the field the method exclusively.

**Comment by JLLeitschuh** [comment]
- [ ] Needs UI Tests

File: ui/src/main/java/edu/wpi/grip/ui/OperationListController.java (hunk 5 above)
Comment (ThomasJClark) [inline ui/src/main/java/edu/wpi/grip/ui/OperationListController.java]: No I  can't, I need to access it from a lambda, and Java doesn't support closures, so it can't just be a local variable

File: ui/src/test/java/edu/wpi/grip/ui/PaletteTest.java [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+        testModule.tearDown();
+    }
+
+    @Test
+    public void testPalette() {
+        // Given a single operation...
+        Operation addition = new AdditionOperation();
+        eventBus.post(new OperationAddedEvent(addition));
+
+        // If we click on the operation button in the palette...
```
Comment (JLLeitschuh) [inline ui/src/test/java/edu/wpi/grip/ui/PaletteTest.java]: I would create an anonymous inner class here instead that did the subscribe.
You should avoid test variables from escaping the system under test.
If other tests are added `step` may not be null because of another test.

The same instance of this test class is reused between every test so you should isolate the output for the test to the scope of the method.

File: ui/src/test/java/edu/wpi/grip/ui/PaletteTest.java [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+
+        FXMLLoader loader = new FXMLLoader(getClass().getResource("Palette.fxml"));
+        loader.setControllerFactory(injector::getInstance);
+        stage.setScene(new Scene(loader.load(), 300, 800));
+        stage.show();
+    }
+
+    @Override
+    public void stop() {
+        testModule.tearDown();
```
Comment (JLLeitschuh) [inline ui/src/test/java/edu/wpi/grip/ui/PaletteTest.java]: You don't set the step to null here so future tests may have step being null. You shouldn't have `step` be scoped so it is visible to the entire class.

File: ui/src/main/java/edu/wpi/grip/ui/OperationListController.java (hunk 5 above)
Comment (JLLeitschuh) [inline ui/src/main/java/edu/wpi/grip/ui/OperationListController.java]: This would work fine:
```java
    @FXML
    public void initialize() {
        operationsMapManager = new ControllerMap<>(operations.getChildren());

        InvalidationListener filterOperations = observable -> {
            final String baseText = root.getText();

            // Show only operations matching the current filter text
            String filter = filterText.getValue();
            long numMatches = operationsMapManager.keySet().stream()
                    .filter(key -> {
                        boolean visible = SearchUtility.fuzzyContains(key.getOperation().getName(), filter)
                                || SearchUtility.fuzzyContains(key.getOperation().getDescription(), filter);
                        operationsMapManager.get(key).setVisible(visible);
                        return visible;
                    }).count();

            if (!filter.isEmpty() && numMatches > 0) {
                // If we're filtering some operations and there's at least one match, set the title to bold and show the
                // number of matches.  This lets the user quickly see which tabs have matching operations when
                // searching.
                root.setText(baseText + " (" + numMatches + ")");
                root.styleProperty().setValue("-fx-font-weight: bold");
            } else {
                root.setText(baseText);
                root.styleProperty().setValue("");
            }
        };

        operations.getChildren().addListener(filterOperations);
        filterText.addListener(filterOperations);

        root.getProperties().addListener((MapChangeListener<? super Object, ? super Object>) change -> {
            if (change.getKey().equals(filterText.getName())) {
                filterText.setValue((String) root.getProperties().get(filterText.getName()));
            }
        });
    }
```

File: ui/src/main/java/edu/wpi/grip/ui/PaletteController.java [hunk 12]
```diff
@@ -58,40 +40,23 @@ public void initialize() {
...
-        operationSearch.textProperty().addListener(filterOperations);
+        imgprocOperations.setUserData(Operation.Category.IMAGE_PROCESSING);
+        featureOperations.setUserData(Operation.Category.FEATURE_DETECTION);
+        networkOperations.setUserData(Operation.Category.NETWORK);
+        opencvOperations.setUserData(Operation.Category.OPENCV);
+        miscellaneousOperations.setUserData(Operation.Category.MISCELLANEOUS);
+
+        // Bind the filterText of all of the individual tabs to the search field
+        operationSearch.textProperty().addListener(observable -> {
+            allOperations.getProperties().put("filterText", operationSearch.getText());
```
Comment (JLLeitschuh) [inline ui/src/main/java/edu/wpi/grip/ui/PaletteController.java]: > Codacy Issue found: [The String literal "filterText" appears 6 times in this file; the first occurrence is on line 57](https://www.codacy.com/app/WPIRoboticsProjects/GRIP/file/2056769135/issues/source?bid=2943107&fileBranchId=3130834#l51)

I agree.

File: ui/src/test/java/edu/wpi/grip/ui/PaletteTest.java [hunk 13]
```diff
@@ -0,0 +1,70 @@
...
+
+public class PaletteTest extends ApplicationTest {
+
+    @Inject private EventBus eventBus;
+
+    private Step step;
+    private final GRIPCoreTestModule testModule = new GRIPCoreTestModule();
+
+    @Override
+    public void start(Stage stage) throws Exception {
```
Comment (JLLeitschuh) [inline ui/src/test/java/edu/wpi/grip/ui/PaletteTest.java]: > Codacy Issue found: [A method/constructor shouldnt explicitly throw java.lang.Exception](https://www.codacy.com/app/WPIRoboticsProjects/GRIP/file/2056769179/issues/source?bid=2943107&fileBranchId=3130834#l34)

It doesn't need to. If there are any exceptions that can be thrown explicitly state them.

Comment (ThomasJClark) [inline ui/src/test/java/edu/wpi/grip/ui/PaletteTest.java]: I can fix it, but for the record it was copied and pasted from `MainWindowTest`.  What was the justification for throwing `Exception` there?

File: ui/src/test/java/edu/wpi/grip/ui/PaletteTest.java (hunk 11 above)
Comment (ThomasJClark) [inline ui/src/test/java/edu/wpi/grip/ui/PaletteTest.java]: k

File: ui/src/test/java/edu/wpi/grip/ui/PaletteTest.java (hunk 10 above)
Comment (ThomasJClark) [inline ui/src/test/java/edu/wpi/grip/ui/PaletteTest.java]: k

File: ui/src/main/java/edu/wpi/grip/ui/PaletteController.java (hunk 12 above)
Comment (ThomasJClark) [inline ui/src/main/java/edu/wpi/grip/ui/PaletteController.java]: k

File: ui/src/main/java/edu/wpi/grip/ui/OperationListController.java (hunk 5 above)
Comment (ThomasJClark) [inline ui/src/main/java/edu/wpi/grip/ui/OperationListController.java]: No it won't.  Just think about that for a minute.

File: ui/src/main/java/edu/wpi/grip/ui/PaletteController.java (hunk 9 above)
Comment (ThomasJClark) [inline ui/src/main/java/edu/wpi/grip/ui/PaletteController.java]: What do you mean commit noise?  I literally have to inject these so I can [set some properties on them](https://github.com/WPIRoboticsProjects/GRIP/pull/502/files#diff-e60dcd0c2e9fa0f698d98ffeda803301R43).

File: ui/src/main/java/edu/wpi/grip/ui/OperationListController.java [hunk 14]
```diff
@@ -0,0 +1,107 @@
...
+/**
+ * Controller for a VBox of {@link OperationController}s.  The user data for this should be what category it shows,
+ * and the filterText property allows for fuzzy searching the operations.
+ *
+ * @see edu.wpi.grip.core.Operation.Category
+ */
+@ParametrizedController(url = "OperationList.fxml")
+public class OperationListController {
+
+    public final static String FILTER_TEXT = "filterText";
```
Comment (JLLeitschuh) [inline ui/src/main/java/edu/wpi/grip/ui/OperationListController.java]: Does this really need to be public? I think protected is enough.

Comment (ThomasJClark) [inline ui/src/main/java/edu/wpi/grip/ui/OperationListController.java]: Yeah, it's part of the public API for this class.

Comment (JLLeitschuh) [inline ui/src/main/java/edu/wpi/grip/ui/OperationListController.java]: Then add documentation explaining what it is for.

File: ui/src/test/java/edu/wpi/grip/ui/PaletteTest.java [hunk 15]
```diff
@@ -0,0 +1,74 @@
...
+import org.testfx.util.WaitForAsyncUtils;
+
+import java.io.IOException;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertNotNull;
+
+public class PaletteTest extends ApplicationTest {
+
+    private final GRIPCoreTestModule testModule = new GRIPCoreTestModule();
```
Comment (JLLeitschuh) [inline ui/src/test/java/edu/wpi/grip/ui/PaletteTest.java]: Don't do this. You need to create a new module for each test.

File: ui/src/main/java/edu/wpi/grip/ui/PaletteController.java (hunk 9 above)
Comment (JLLeitschuh) [inline ui/src/main/java/edu/wpi/grip/ui/PaletteController.java]: Sure, fine.

File: ui/src/main/java/edu/wpi/grip/ui/OperationListController.java [hunk 16]
```diff
@@ -0,0 +1,105 @@
...
+
+    @FXML private Tab root;
+    @FXML private VBox operations;
+
+    @Inject private OperationController.Factory operationControllerFactory;
+
+    private final StringProperty filterText = new SimpleStringProperty(this, FILTER_TEXT, "");
+    private String baseText = null;
```
Comment (JLLeitschuh) [inline ui/src/main/java/edu/wpi/grip/ui/OperationListController.java]: The robot is right, this doesn't need to be a class variable.
See this: https://github.com/WPIRoboticsProjects/GRIP/pull/502#discussion_r53109575

Comment (ThomasJClark) [inline ui/src/main/java/edu/wpi/grip/ui/OperationListController.java]: No, I already told you  **that doesn't work**.  Did you even try running that code?  It keeps getting the current text in the title and setting it to `baseText`.  `baseText` is supposed to have just the title text without the parentheses and number.  Retrieving it over and over again means you'll have a bunch of those at the end of the title (like "Image processing (1) (2) (3) (4) (5)...".

File: ui/src/main/java/edu/wpi/grip/ui/OperationListController.java [hunk 17]
```diff
@@ -0,0 +1,105 @@
...
+        filterText.addListener(filterOperations);
+
+        root.getProperties().addListener((MapChangeListener<? super Object, ? super Object>) change -> {
+            if (change.getKey().equals(filterText.getName())) {
+                filterText.setValue((String) root.getProperties().get(filterText.getName()));
+            }
+        });
+    }
+
+    public StringProperty filterTextProperty() {
```
Comment (JLLeitschuh) [inline ui/src/main/java/edu/wpi/grip/ui/OperationListController.java]: Where is this method even used? From within FXML? Can you mark it that way?
